### ISSUE TITLE
[icetray] I3Frame::Get for shared_ptrs always throws an exception when the object cannot be deserialized (Trac #1570)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#1562]
I3Frame's internal `get_impl` throws an `archive_exception` when asked to fetch an object which cannot be deserialized. The reference version of `Get`can't do much in this case and should just propagate the exception, but the `shared_ptr` version could return a null pointer to indicate the lack of a valid object of the right type. In fact, it already has the `quietly` flag which suppresses printing a message in this case, so it would make sense for it to also handle the exception, at least if this flag is set. If the flag is not set we can let the user do whatever he or she wants about the exception, as is currently required. 

Patch implementing the proposed change:

```text
Index: public/icetray/I3Frame.h
===================================================================
--- public/icetray/I3Frame.h	(revision 2423)
+++ public/icetray/I3Frame.h	(working copy)
@@ -29,6 +29,7 @@
 #include <boost/iterator/transform_iterator.hpp>
 #include <boost/utility/enable_if.hpp>
 #include <boost/type_traits/is_const.hpp>
+#include <boost/archive/archive_exception.hpp>
 
 #include <icetray/I3DefaultName.h>
 #include <icetray/I3FrameObject.h>
@@ -307,7 +308,8 @@
    * object...  e.g. there might be multiple I3Particles stored at
    * different names like "linefit" and "dipolefit", and there might
    * be only one I3Geometry stored at "I3Geometry".
-   * @param quietly Be quiet or not
+   * @param quietly Whether a non-deserializable frame object is treated as an 
+   * error or not.
    * @return shared_ptr<const T::value_type>.  That is, if Get is
    * called as Get<shared_ptr<const T> >(name), or the equivalent
    * Get<I3SomethingConstPtr>(name), this function returns
@@ -315,6 +317,8 @@
    * shared_ptr, const is added: e.g. Get<I3SomethingPtr>(name)
    * returns I3SomethingConstPtr.
    * More generally, Get<shared_ptr<T> >(name) returns shared_ptr<const T>.
+   * @throw If quietly is false and the frame object cannot be deserialized, 
+   * a boost::archive::archive_exception will be thrown. 
    */
   template <typename T>
   T
@@ -329,7 +333,18 @@
     if (iter == map_.end())
       return boost::shared_ptr<typename T::element_type>();
 
-    I3FrameObjectConstPtr focp = get_impl(*iter, quietly);
+    I3FrameObjectConstPtr focp;
+    try{
+      focp = get_impl(*iter, quietly);
+    } catch (const boost::archive::archive_exception& e) {
+      //This will happen if the the frame object could not be deserialized.
+      //If we were asked to run quietly, just treat it as not being a valid
+      //instance of T::element_type, and return a null pointer.
+      if (quietly)
+        return T();
+      //otherwise, propagate the exception to the user
+      throw;
+    }
     
     return boost::dynamic_pointer_cast<typename T::element_type>(focp);
   }
```

Migrated from https://code.icecube.wisc.edu/ticket/1570
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:13:57", 
    "description": "I3Frame's internal `get_impl` throws an `archive_exception` when asked to fetch an object which cannot be deserialized. The reference version of `Get`can't do much in this case and should just propagate the exception, but the `shared_ptr` version could return a null pointer to indicate the lack of a valid object of the right type. In fact, it already has the `quietly` flag which suppresses printing a message in this case, so it would make sense for it to also handle the exception, at least if this flag is set. If the flag is not set we can let the user do whatever he or she wants about the exception, as is currently required. \n\nPatch implementing the proposed change:\n\n{{{\nIndex: public/icetray/I3Frame.h\n===================================================================\n--- public/icetray/I3Frame.h\t(revision 2423)\n+++ public/icetray/I3Frame.h\t(working copy)\n@@ -29,6 +29,7 @@\n #include <boost/iterator/transform_iterator.hpp>\n #include <boost/utility/enable_if.hpp>\n #include <boost/type_traits/is_const.hpp>\n+#include <boost/archive/archive_exception.hpp>\n \n #include <icetray/I3DefaultName.h>\n #include <icetray/I3FrameObject.h>\n@@ -307,7 +308,8 @@\n    * object...  e.g. there might be multiple I3Particles stored at\n    * different names like \"linefit\" and \"dipolefit\", and there might\n    * be only one I3Geometry stored at \"I3Geometry\".\n-   * @param quietly Be quiet or not\n+   * @param quietly Whether a non-deserializable frame object is treated as an \n+   * error or not.\n    * @return shared_ptr<const T::value_type>.  That is, if Get is\n    * called as Get<shared_ptr<const T> >(name), or the equivalent\n    * Get<I3SomethingConstPtr>(name), this function returns\n@@ -315,6 +317,8 @@\n    * shared_ptr, const is added: e.g. Get<I3SomethingPtr>(name)\n    * returns I3SomethingConstPtr.\n    * More generally, Get<shared_ptr<T> >(name) returns shared_ptr<const T>.\n+   * @throw If quietly is false and the frame object cannot be deserialized, \n+   * a boost::archive::archive_exception will be thrown. \n    */\n   template <typename T>\n   T\n@@ -329,7 +333,18 @@\n     if (iter == map_.end())\n       return boost::shared_ptr<typename T::element_type>();\n \n-    I3FrameObjectConstPtr focp = get_impl(*iter, quietly);\n+    I3FrameObjectConstPtr focp;\n+    try{\n+      focp = get_impl(*iter, quietly);\n+    } catch (const boost::archive::archive_exception& e) {\n+      //This will happen if the the frame object could not be deserialized.\n+      //If we were asked to run quietly, just treat it as not being a valid\n+      //instance of T::element_type, and return a null pointer.\n+      if (quietly)\n+        return T();\n+      //otherwise, propagate the exception to the user\n+      throw;\n+    }\n     \n     return boost::dynamic_pointer_cast<typename T::element_type>(focp);\n   }\n}}}", 
    "reporter": "cweaver", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1550067237750774", 
    "component": "combo core", 
    "summary": "[icetray] I3Frame::Get for shared_ptrs always throws an exception when the object cannot be deserialized", 
    "priority": "critical", 
    "keywords": "", 
    "time": "2016-02-26T18:11:20", 
    "milestone": "", 
    "owner": "olivas", 
    "type": "enhancement"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20160229T19:31:38: 
* olivas changed owner from "cweaver" to "olivas"
* olivas changed status from "new" to "accepted"

**Comment by icecube-trac** [comment]
Trac update at 20160229T19:59:41: olivas commented:

> As is, this would break dataio-pyshovel, which catches the exception, which is how it determines that objects are 'Unreadable.'
> 
> http://code.icecube.wisc.edu/projects/icecube/browser/IceTray/projects/dataio/trunk/python/pprint.py#L27
> 
> It turns out that __getitem__ is bound to frame_get : http://code.icecube.wisc.edu/projects/icecube/browser/IceTray/projects/icetray/trunk/private/pybindings/I3Frame.cxx#L28
> 
> ...which has quietly set to true.
> 
> It might be as easy as just setting that to false *or* instead of catching the exception in the pybindings we check for a non-NULL pointer.  I'm preferring the latter. The former would spew and throw, which is only half of what we want.  The latter still suppresses the spew and checking for a NULL pointer is cleaner than a try-catch.
> 
> Might be worth checking steamshovel, but I'm now wondering what other downstream code might depend on this behavior.

**Comment by icecube-trac** [comment]
Trac update at 20160229T22:06:47: olivas commented:

> Now that I think about it "quietly" should have never existed in the first place.  If it's really OK for that log_error to be quiet it should be a log_debug or log_trace.  It's really that simple.

**Comment by icecube-trac** [comment]
Trac update at 20160229T22:10:26: olivas commented:

> Now the real question is, should getting by pointer throw if an object can't be deserialized?
> 
> On one hand, I actually like that in any case it throws with "Frame caught exception..." That's a pretty clear indication (at least to some of us) as to what the problem is.
> 
> Although I3Frame::Has and NOT is pretty much all you need to narrow it down too.

**Comment by icecube-trac** [comment]
Trac update at 20160426T15:05:19: olivas changed milestone from "" to "August 2016"

**Comment by icecube-trac** [comment]
Trac update at 20170118T19:14:17: olivas changed summary from "I3Frame::Get for shared_ptrs always throws an exception when the object cannot be deserialized" to "[icetray] I3Frame::Get for shared_ptrs always throws an exception when the object cannot be deserialized"

**Comment by icecube-trac** [comment]
Trac update at 20170124T00:59:56: 
* olivas changed resolution from "" to "fixed"
* olivas changed status from "accepted" to "closed"
* olivas commented:

> Patch accepted in r2774.

**Comment by icecube-trac** [comment]
Trac update at 20170124T23:15:01: olivas commented:

> I've had some time to think and I'm pretty set on the current C++ implementation and more convinced the old implementation was just bad design.
> 
> We have two forms of Get.  One returns a pointer and one returns a reference.  The one that returns a reference *has* to throw.  There's no other option.  Because of this I think it's reasonable that the one that returns a shared pointer never throw.  After all it's perfectly natural for *any* function/method that returns an object by pointer, to return a NULL (or the equivalent) when things go wrong.  This is just what people expect, so the semantics are natural.  Not only is it good coding practice, but it should almost be considered law that if you call a method that returns a pointer to an object, YOU MUST check that it's valid before dereferencing it.  You also must therefore write code to handle the case that it is NULL.  So you either have an if-then-else or a try-catch, depending on whether the method that returns a pointer or a reference.  Allowing the method that returns a pointer to sometimes throw sometimes, sometimes spew and sometimes not depending on a flag is just clunky and unnecessary, in my opinion.
> 
> The question now then is what should happen on the python-side.  I'm not opposed to restoring the old semantics, but we can do this in the bindings, keeping the current C++ implementation.  I prefer it to return None, but I'm open to suggestions.

**Comment by icecube-trac** [comment]
Trac update at 20170125T00:36:55: david.schultz commented:

> On the python side, raising an Exception is the correct syntax if anything goes wrong. `None` really means nothing is there, not that there was a problem.  Also note that python doesn't have pointers in the language, everything is references.

**Comment by icecube-trac** [comment]
Trac update at 20170310T21:32:54: 
* david.schultz changed resolution from "fixed" to ""
* david.schultz changed status from "closed" to "reopened"

**Comment by icecube-trac** [comment]
Trac update at 20170418T17:26:35: david.schultz commented:

> Here's the diff, if anyone's interested.  This makes python revert to the old behavior where it gets an Exception.
> 
> 
> ```text
> Index: private/pybindings/I3Frame.cxx
> ===================================================================
> --- private/pybindings/I3Frame.cxx	(revision 2855)
> +++ private/pybindings/I3Frame.cxx	(working copy)
> @@ -37,6 +37,12 @@
>      }
>  
>    boost::shared_ptr<const T> thing = f->Get<boost::shared_ptr<const T> >(where);
> +  if (!thing)
> +    {
> +      PyErr_SetString(PyExc_RuntimeError, "Serialization error");
> +      throw_error_already_set();
> +      return boost::shared_ptr<T>();
> +    }
>    return boost::const_pointer_cast<T>(thing);
>  }
> ```

**Comment by icecube-trac** [comment]
Trac update at 20170418T17:33:10: 
* olivas changed milestone from "August 2016" to "May 2017"
* olivas commented:

> Because this could change downstream scripts, gonna leave this for the next release, as opposed to trying to get it into IceRec 5.1.  Filtering works.

**Comment by icecube-trac** [comment]
Trac update at 20170425T16:43:41: david.schultz commented:

> Note that I accidentally didn't load dataclasses today, and got this strange effect:
> 
> 
> ```text
> In [7]: fr['I3EventHeader']
> Out[7]: <icecube.icetray.I3FrameObject at 0x7f91eec115f0>
> 
> In [8]: e= fr['I3EventHeader']
> 
> In [9]: print e
> <icecube.icetray.I3FrameObject object at 0x7f91eec11668>
> 
> In [10]: from icecube import dataclasses
> 
> In [11]: fr['I3EventHeader']
> Out[11]: <icecube.dataclasses.I3EventHeader at 0x7f91ee5b0848>
> ```
> 
> So returning an empty I3FrameObject to python is really not nice.  You can't compare it to anything to test for bad serialization.

**Comment by icecube-trac** [comment]
Trac update at 20170426T22:18:56: olivas commented:

> That's certainly an excellent reason to throw.

**Comment by icecube-trac** [comment]
Trac update at 20170430T21:16:20: david.schultz commented:

> So apparently the really interesting situation that I can't figure out how to catch is when you've imported dataio, but not dataclasses.  This is because dataio links to dataclasses, so it can deserialize it but not convert to python (thus remains as an `I3FrameObject`).

**Comment by icecube-trac** [comment]
Trac update at 20170503T20:51:32: olivas commented:

> this is an easy fix in dataio's __init__.py, no?  i think we've done this with other projects.

**Comment by icecube-trac** [comment]
Trac update at 20170503T20:55:57: olivas changed status from "reopened" to "accepted"

**Comment by icecube-trac** [comment]
Trac update at 20170516T22:08:15: david.schultz commented:

> r2869/IceTray changed dataio's `__init__.py`, fixing my "interesting situation".
> 
> So now we just have whether to raise or not.

**Comment by icecube-trac** [comment]
Trac update at 20170627T16:00:40: 
* david.schultz commented:

> Note that a user got bitten by this.  From slack:
> 
> wardvandriessche:
> This might be a very stupid question, but does anybody know why I see 'MCPESeriesMap' in dataio-pyshovel but when I try to access it in python I get None?
> 
> 
> ```text
> In [2]: from icecube import dataclasses,dataio
> 
> In [3]: file = dataio.I3File('SMP_Charge_0.333333333333_Mass_10.0GeV_n_Events_500_File_0_Level0_beforeTrig.i3.bz2')
> 
> In [4]: ph = file.pop_daq()
> 
> In [5]: print ph['MCPESeriesMap']
> None
> ```
> 


* david.schultz changed priority from "normal" to "critical"

**Comment by icecube-trac** [comment]
Trac update at 20170627T16:04:03: olivas commented:

> To be fair the question would have changed to "Why did this throw?" and our answer would have been the same.  In any case, this is an issue I want to close this week and it should only take me a solid hour.  The tweaks to icetray should be minimal, but the shovels need to be tweaked accordingly.
> 
> Also even in the current implementation of dataio-pyshovel that should show up as "Unreadable."

**Comment by icecube-trac** [comment]
Trac update at 20170627T16:06:31: david.schultz commented:

> The shovels read it properly because they import dataclasses, simclasses, and recclasses.
> 
> I like raising an exception because you can give it a string error message, such as "serialization failed - maybe you forgot to import a library?".

**Comment by icecube-trac** [comment]
Trac update at 20170627T16:10:07: olivas commented:

> Replying to [comment:22 david.schultz]:
> > The shovels read it properly because they import dataclasses, simclasses, and recclasses.
> > 
> > I like raising an exception because you can give it a string error message, such as "serialization failed - maybe you forgot to import a library?".
> 
> Yup, although we could still emit a warning in that case.  If the object exists in the frame and a null pointer is returned then log_error("...").

**Comment by icecube-trac** [comment]
Trac update at 20170627T16:15:20: david.schultz commented:

> Note that it's not fun to handle log messages with dataio-pyshovel, because it's drawing over the console.  The log message generally interrupts the current frame, and only half the time the next draw call fixes it.  Most times it gets offset somehow.
> 
> I'd also argue against log messages, because most people ignore them a lot.  We have a history of so many log messages in real processing that people ignore anything less than a fatal. (they'll go back and look if something fatals, but otherwise ignore)

**Comment by icecube-trac** [comment]
Trac update at 20170627T16:22:39: olivas commented:

> The first point is fair.  The second...eh...we might as well just not log at all in that case. :)  We could greatly simplify our lives if we only supported log_fatal.

**Comment by icecube-trac** [comment]
Trac update at 20170627T16:26:07: david.schultz commented:

> My point is that logging is probably only useful once a fatal problem has occurred. So when you're debugging, it's useful.  When actually running in release, it's not useful at all.
> 
> The one thing I am trying is to raise the default log level in dataio-pyshovel to fatal, so all other log messages are suppressed.

**Comment by icecube-trac** [comment]
Trac update at 20170627T16:33:49: olivas commented:

> So if most logging is suppressed in dataio-pyshovel then a log_error (or _warn) doesn't hurt, and it looks like you have to do this anyway.  
> 
> Ward would have caught this in his python session, so problem solved there.
> 
> In production do we really want to shut things down because someone forgot to load a library?  If not and we're just going to catch that anyways, then why throw in the first place?

**Comment by icecube-trac** [comment]
Trac update at 20170627T16:36:57: david.schultz commented:

> My problem is that returning a None allows a script to continue.  It's likely to fail at some point, but that may not be for hundreds of lines in some other obscure part of the code, leading to the user chasing down the wrong direction.
> 
> If the user explicitly adds a try block, then they know what they are doing and are prepared to handle the bad serialization.

**Comment by icecube-trac** [comment]
Trac update at 20170627T17:22:29: olivas commented:

> You can continue in either fashion and returning a None is just as likely to raise an exception as soon as you try to use it, so pretty early.
> 
> In both cases the user has to decide what to do in case of failure.  They either check the object is not None or they catch the throw.

**Comment by icecube-trac** [comment]
Trac update at 20170627T17:22:55: olivas commented:

> Don't worry, I'm still going to change that to throw. :)

**Comment by icecube-trac** [comment]
Trac update at 20170627T23:43:35: 
* olivas commented:

> In changeset 2924:
> 
> ```text
> #!CommitTicketReference repository="IceTray" revision="2924"
> throw in python with a nice error message if deserialization fails.  closes #1562.
> ```


* olivas changed status from "accepted" to "closed"
* olivas changed resolution from "" to "fixed"

**Comment by icecube-trac** [comment]
Trac update at 20190111T20:59:41: nwhitehorn commented:

> In changeset 168823:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="168823"
> throw in python with a nice error message if deserialization fails.  closes #1562.
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:13:57: 
* olivas changed milestone from "June 2017" to ""
* olivas commented:

> Milestone June 2017 deleted
